pkg/cli/admin/release/info.go

```diff
@@ -1987,6 +2035,306 @@ func describeBugs(out, errOut io.Writer, diff *ReleaseDiff, dir string, format s
+		return err
+	}
+	if toRef, err = imagereference.Parse(osDiff.To.From.Name); err != nil {
+		return err
+	}
+
+	// not sure this can ever happen, but just sanity-check anyway
+	if fromRef.ID == "" {
+		return fmt.Errorf("image not digest-based: %s", fromRef)
+	}
+	if toRef.ID == "" {
+		return fmt.Errorf("image not digest-based: %s", toRef)
+	}
+
+	fromDbPath := filepath.Join(dir, fromRef.ID)
+	if _, err = os.Stat(fromDbPath); os.IsNotExist(err) {
+		err = o.downloadRpmdb(fromRef, fromDbPath)
+	}
+	if err != nil {
+		return err
+	}
+
+	toDbPath := filepath.Join(dir, toRef.ID)
+	if _, err := os.Stat(toDbPath); os.IsNotExist(err) {
+		err = o.downloadRpmdb(toRef, toDbPath)
+	}
+	if err != nil {
+		return err
+	}
+
+	fromDb, err := loadRpmdb(fromDbPath)
+	if err != nil {
+		return err
+	}
+
+	toDb, err := loadRpmdb(toDbPath)
+	if err != nil {
+		return err
+	}
+
+	removed := make(map[string]string)
+	changed := make(map[string]RpmChangedDiff)
+	for pkg, v := range fromDb {
+		if newVersion, ok := toDb[pkg]; ok {
+			if v != newVersion {
+				changed[pkg] = RpmChangedDiff{
+					Old: v,
+					New: newVersion,
+				}
+			}
+			delete(toDb, pkg)
+		} else {
+			removed[pkg] = v
+		}
+	}
+	rpmDiff := RpmDiff{
+		Changed: changed,
+		Added:   toDb,
+		Removed: removed,
+	}
+
+	if format == "json" {
+		data, err := json.MarshalIndent(rpmDiff, "", "  ")
+		if err != nil {
+			return err
+		}
+		fmt.Fprintln(o.Out, string(data))
+	} else {
+		writeList := func(header string, elements []string) {
+			fmt.Fprintln(o.Out, header)
+			sort.Strings(elements)
+			for _, elem := range elements {
+				fmt.Fprintln(o.Out, elem)
+			}
+		}
+
+		if len(rpmDiff.Changed) > 0 {
+			elements := []string{}
+			for pkg, v := range rpmDiff.Changed {
+				elements = append(elements, fmt.Sprintf("  %s %s → %s", pkg, v.Old, v.New))
+			}
+			writeList("Changed:", elements)
+		}
+		if len(rpmDiff.Removed) > 0 {
+			elements := []string{}
+			for pkg, v := range rpmDiff.Removed {
+				elements = append(elements, fmt.Sprintf("  %s %s", pkg, v))
+			}
+			writeList("Removed:", elements)
+		}
+		if len(rpmDiff.Added) > 0 {
+			elements := []string{}
+			for pkg, v := range rpmDiff.Added {
+				elements = append(elements, fmt.Sprintf("  %s %s", pkg, v))
+			}
+			writeList("Added:", elements)
+		}
+	}
+
+	return nil
+}
+
+type RpmDiff struct {
+	Changed map[string]RpmChangedDiff `json:"changed,omitempty"`
+	Added   map[string]string         `json:"added,omitempty"`
+	Removed map[string]string         `json:"removed,omitempty"`
+}
+
+type RpmChangedDiff struct {
+	Old string `json:"old,omitempty"`
+	New string `json:"new,omitempty"`
+}
+
+func (o *InfoOptions) downloadRpmdb(image imagereference.DockerImageReference, dir string) error {
+	// NOTE: in the future, should check if there's an SBOM and get the rpmlist from that instead
+
+	ref := imagesource.TypedImageReference{Type: imagesource.DestinationRegistry, Ref: image}
+
+	tmpDir := dir + ".work"
```

Comment:
Can you use [`os.MkdirTemp`][1] here?  Or are you worried about leaking temporary directories if the process is terminated mid-download?  There's some tension here between "the release-controller use-case is one long-running process, and we'd rather not leak directories there" and "`oc adm release info ...` can be used in other places, with multiple invocations in parallel, and if we hard-code this kind of scratch directory, we could have two parallel runs corrupting the dir they didn't realize they were sharing".  I guess also theoretically on the table would be lock files or some such so at least parallel invocations could notice overlap and exit out?  With some way to have the lock expire to recover from died-mid-download?  Lots of corner cases...

[1]: https://pkg.go.dev/os#MkdirTemp

Reply:
Hmm, yes good point. I need to check what the release controller does if multiple requests come in for the same thing. This must be a solved case at least for the git-based stuff which already exists. Overall, it feels like locking/"request joining" (for lack of a better term; i.e. having a secondary request join an earlier matching request already ongoing) should live at the release controller level (or similarly anything driving potentially multiple invocations of `oc`) rather than within `oc` using file locks?

Reply:
release-controller isn't the only `oc` caller though.  It's theoretically possible that a user like me with my own `oc` copy runs this command twice in the same directory at the same time

Reply:
The release-controller runs its `oc adm release info` commands in memory now (not out on its own pod).  It will cache results based on the `from` and `to` versions being compared.

Reply:
> release-controller isn't the only `oc` caller though. It's theoretically possible that a user like me with my own `oc` copy runs this command twice in the same directory at the same time

I'm not too worried about the user case TBH, though OTOH you're right that if locking needs to be added somewhere, it might as well be here to cover all possible callers. Added basic locking for now! We can make it fancier down the road if needed.